File: 4.Programming/lovemoney_fe/lib/features/domain/entities/transaction.dart
class Transaction {
  final String? _name;
  final double? _cost;
  final String? _type;
  final String? _typeTransaction;
  final String? _note;

  Transaction(
      {String? name,
      double? cost,
      String? type,
      String? typeTransaction,
      String? note})
      : _name = name,
        _cost = cost,
        _type = type,
        _typeTransaction = typeTransaction,
        _note = note;

  Transaction copyWith(
      {String? name,
      double? cost,
      String? type,
      String? typeTransaction,
      String? note}) {
    return Transaction(
      name: name ?? _name,
      cost: cost ?? _cost,
      type: type ?? _type,
      typeTransaction: typeTransaction ?? _typeTransaction,
      note: note ?? _note,
    );
  }
}

class FixedTransaction extends Transaction {
  final double? _periodTime;

  FixedTransaction(
      {String? name,
      double? cost,
      String? type,
      String? typeTransaction,
      String? note,
      double? periodTime})
      : _periodTime = periodTime,
        super(
          name: name,
          cost: cost,
          type: type,
          typeTransaction: typeTransaction,
          note: note,
        );


  @override
  FixedTransaction copyWith({String? name,
      double? cost,
      String? type,
      String? typeTransaction,
      String? note,
      double? periodTime,}) {
    return FixedTransaction(
      name: name ?? _name,
      cost: cost ?? _cost,
      type: type ?? _type,
      typeTransaction: typeTransaction ?? _typeTransaction,
      note: note ?? _note,
      periodTime: periodTime ?? _periodTime,
    );
  }
}

File: 4.Programming/lovemoney_be/models/transaction.js
const mongoose = require('mongoose');
const {TYPE_TRANSACTION, TYPE_TRANSACTION_PART} = require('../utils/constants');
const {ObjectId} = mongoose.Types;

const fixedTransaction = mongoose.Schema({
  periodTime: {
    type: Number,
    require: true,
  },
});

const variableTransaction = new mongoose.Schema({

});

const transactionPartSchema = new mongoose.Schema({
  typeTransactionPart: {
    type: String,
    require: true,
    enum: [
      TYPE_TRANSACTION_PART.FIXED_TRANSACTION,
      TYPE_TRANSACTION_PART.VARIABLE_TRANSACTION,
    ]
  },
  fixedTransaction, 
  variableTransaction, 
});

const transactionSchema = new mongoose.Schema({
  userId: {
    //type: ObjectId,
    type: String,
    ref: 'User',
  },
  name: String,
  cost: Number,
  date: {
    type: Date,
  },
  typeTransaction: {
    type: String,
    enum: [
      TYPE_TRANSACTION.EXPENSE_TRANSACTION,
      TYPE_TRANSACTION.TURNOVER_TRANSACTION,
    ],
  },
  transactionPart: transactionPartSchema, 
  note: String,
})

module.exports = mongoose.model('Transaction', transactionSchema);